test_util/src/runtime.rs

```diff
@@ -0,0 +1,209 @@
+use futures::FutureExt;
+use janus::Runtime;
+use std::{
+    collections::HashMap,
+    future::Future,
+    hash::Hash,
+    panic::{panic_any, AssertUnwindSafe},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+use tokio::{
+    sync::watch::{self, Sender},
+    task::JoinHandle,
+};
+
+/// Tracks multiple instrumented [`Runtime`] objects, for use in tests. Each
+/// [`TestRuntime`] keeps track of how many of its tasks have been completed,
+/// and tests can wait until a given number of tasks finish. If any task
+/// panics, this manager object will panic on drop, to ensure that the
+/// relevant test fails and the task panics do not go unnoticed.
+pub struct TestRuntimeManager<L> {
+    map: HashMap<L, TestRuntime>,
+}
+
+impl<L> TestRuntimeManager<L> {
+    pub fn new() -> TestRuntimeManager<L> {
+        TestRuntimeManager {
+            map: HashMap::new(),
+        }
+    }
+}
+
+impl<L> TestRuntimeManager<L>
+where
+    L: Eq + Hash,
+{
+    /// Construct or retrieve a [`TestRuntime`] with a given label.
+    pub fn with_label(&mut self, label: L) -> TestRuntime {
+        self.map.entry(label).or_default().clone()
+    }
+
+    /// Wait for the runtime with the given label to reach some number of
+    /// completed tasks. Note that tasks that have already completed before
+    /// this method is called are included in the count.
+    pub async fn wait_for_completed_tasks(&mut self, label: L, target_count: usize) {
+        let labeled_runtime = self.with_label(label);
+        let mut receiver = labeled_runtime.inner.sender.subscribe();
+        if *receiver.borrow_and_update() >= target_count {
```

Comment:
nit: If I'm not missing something, this [L50-61] could be slightly simplified:

```rust
loop {
  if *receiver.borrow_and_update() >= target_count { break; }
  receiver.changed.await.expect("...");
}
```

(this may mark values as seen twice, but I think the code simplification is worth it)